pkg/mcp/sink/journal.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sink
+
+import (
+	"sync"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+
+	mcp "istio.io/api/mcp/v1alpha1"
+)
+
+// RecentRequestInfo is metadata about a request that the client has sent.
+type RecentRequestInfo struct {
+	Time    time.Time
+	Request *mcp.RequestResources
+}
+
+// Acked indicates whether the message was an ack or not.
+func (r RecentRequestInfo) Acked() bool {
+	return r.Request.ErrorDetail == nil
+}
+
+const journalDepth = 20
+
+// recentRequestsJournal captures debug metadata about the latest requests that was sent by this client.
+type recentRequestsJournal struct {
+	itemsMutex sync.Mutex
+	items      []RecentRequestInfo
+}
+
+func (r *recentRequestsJournal) record(req *mcp.RequestResources) { // nolint:interfacer
+	r.itemsMutex.Lock()
+	defer r.itemsMutex.Unlock()
+
+	item := RecentRequestInfo{
+		Time:    time.Now(),
+		Request: proto.Clone(req).(*mcp.RequestResources),
+	}
+
+	r.items = append(r.items, item)
+	for len(r.items) > journalDepth {
+		r.items = r.items[1:]
```

Comment:
Is this really what you want here? I think this will end up allocating an ever-increasing array under the covers.

Reply:
This was copy/pasted from the original MCP journal code. I think the same concern was raised during the review of that code, but @ozevren said the golang runtime would reclaim the memory. In any case, I rewrote to use a fixed slice without additional memory allocations.